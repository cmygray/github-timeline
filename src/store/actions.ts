import { ActionTree } from 'vuex';

import { RootState } from '@/store/index';
import {
  Repository,
  UserRepositoriesData,
  UserRepositoriesVariable,
} from '@/models';
import { client, USER_REPOSITORIES_QUERY } from '@/apis';

export const actions: ActionTree<RootState, RootState> = {
  async fetchRepositories({ dispatch }, { userId }): Promise<Repository[]> {
    const { data } = await client.query<
      /* eslint-disable @typescript-eslint/indent */
      UserRepositoriesData,
      UserRepositoriesVariable
      /* eslint-enable @typescript-eslint/indent */
    >({
      variables: { userId },
      query: USER_REPOSITORIES_QUERY,
    });

    if (!data.user) return [];

    return dispatch('setRepositories', data);
  },

  async setRepositories(
    { commit },
    { user: { login, repositories } }
  ): Promise<Repository[]> {
    commit('SET_REPOSITORIES', {
      ownerId: login,
      repositories: repositories.nodes,
    });

    return repositories.nodes;
  },
};
