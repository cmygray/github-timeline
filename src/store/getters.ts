import { GetterTree } from 'vuex';

import { RootState } from '@/store/index';
import { Repository } from '@/models';

export const getters: GetterTree<RootState, RootState> = {
  repositoriesByOwnerId: (
    state: RootState
  ): ((ownerId: string) => Repository[]) => (ownerId: string): Repository[] => {
    return state.repositories[ownerId] || [];
  },
};
