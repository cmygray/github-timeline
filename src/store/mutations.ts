import Vue from 'vue';
import { MutationTree } from 'vuex';

import { RootState } from '@/store/index';

export const mutations: MutationTree<RootState> = {
  SET_REPOSITORIES(state, { ownerId, repositories }): void {
    Vue.set(state.repositories, ownerId, repositories);
  },

  SET_PAGINATION(state, { totalCount, pageInfo }): void {
    const { endCursor, hasNextPage } = pageInfo;

    Vue.set(state, 'pagination', {
      totalCount,
      endCursor,
      hasNextPage,
    });
  },
};
