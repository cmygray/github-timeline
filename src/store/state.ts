import { RootState } from '@/store/index';

export const state: RootState = {
  repositories: {},
};
