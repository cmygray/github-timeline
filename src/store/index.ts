import Vue from 'vue';
import Vuex, { Store } from 'vuex';
import createLogger from 'vuex/dist/logger';

import { Repository } from '@/models';
import { state } from '@/store/state';
import { getters } from '@/store/getters';
import { mutations } from '@/store/mutations';
import { actions } from '@/store/actions';

Vue.use(Vuex);

const debug = process.env.NODE_ENV !== 'production';

export interface RootState {
  pagination: {
    totalCount?: number;
    endCursor?: string;
    hasNextPage?: boolean;
  };
  repositories: {
    [ownerId: string]: Repository[];
  };
}

export function createStore(): Store<RootState> {
  return new Vuex.Store<RootState>({
    state,
    getters,
    mutations,
    actions,
    plugins: debug ? [createLogger()] : [],
  });
}
