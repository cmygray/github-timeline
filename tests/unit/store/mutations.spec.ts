import { mutations } from '@/store/mutations';

import { repositories } from '../__fixtures__/repositories';
import { RootState } from '@/store';

const state: RootState = { repositories: {} };

describe('mutations', () => {
  describe('SET_REPOSITORIES', () => {
    const PAYLOAD = {
      ownerId: 'cmygray',
      repositories: repositories
    };
    it('sets repositories of an owner', () => {
      mutations.SET_REPOSITORIES(state, PAYLOAD);

      expect(state).toMatchSnapshot();
      expect(state.repositories[PAYLOAD.ownerId]).toEqual(PAYLOAD.repositories)
    });
  })
});
