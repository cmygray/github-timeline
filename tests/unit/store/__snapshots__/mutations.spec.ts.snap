// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mutations SET_REPOSITORIES sets repositories of an owner 1`] = `
Object {
  "repositories": Object {
    "cmygray": Array [
      Object {
        "createdAt": "2019-06-10T08:44:44Z",
        "description": "Create GraphQL schema and resolvers with TypeScript, using classes and decorators!",
        "id": "MDEwOlJlcG9zaXRvcnkxOTExMzA0MTQ=",
        "isFork": true,
        "name": "type-graphql",
        "stargazers": Object {
          "totalCount": 0,
        },
        "updatedAt": "2019-06-10T08:44:47Z",
      },
      Object {
        "createdAt": "2019-06-09T07:19:12Z",
        "description": null,
        "id": "MDEwOlJlcG9zaXRvcnkxOTA5NzYxMzg=",
        "isFork": false,
        "name": "github-timeline",
        "stargazers": Object {
          "totalCount": 0,
        },
        "updatedAt": "2019-06-10T04:54:46Z",
      },
      Object {
        "createdAt": "2019-06-07T07:53:37Z",
        "description": "A boilerplate for Vue.js using typescript, jest, router, Vuex...",
        "id": "MDEwOlJlcG9zaXRvcnkxOTA3MDc1MjI=",
        "isFork": false,
        "name": "vue-boilerplate",
        "stargazers": Object {
          "totalCount": 0,
        },
        "updatedAt": "2019-06-09T06:20:56Z",
      },
      Object {
        "createdAt": "2019-04-25T11:06:37Z",
        "description": "Simple Vue.js authentication library",
        "id": "MDEwOlJlcG9zaXRvcnkxODM0MTczODY=",
        "isFork": true,
        "name": "vue-authenticate",
        "stargazers": Object {
          "totalCount": 0,
        },
        "updatedAt": "2019-04-25T11:06:40Z",
      },
      Object {
        "createdAt": "2019-04-18T12:42:29Z",
        "description": "Delightful JavaScript Testing.",
        "id": "MDEwOlJlcG9zaXRvcnkxODIwODU4MTI=",
        "isFork": true,
        "name": "jest",
        "stargazers": Object {
          "totalCount": 0,
        },
        "updatedAt": "2019-04-18T12:42:35Z",
      },
      Object {
        "createdAt": "2019-04-02T05:16:18Z",
        "description": null,
        "id": "MDEwOlJlcG9zaXRvcnkxNzkwMDM3MzY=",
        "isFork": true,
        "name": "reststate-client",
        "stargazers": Object {
          "totalCount": 0,
        },
        "updatedAt": "2019-04-04T02:09:26Z",
      },
      Object {
        "createdAt": "2019-02-11T09:05:16Z",
        "description": "Model Driven State Management",
        "id": "MDEwOlJlcG9zaXRvcnkxNzAwOTg1NjY=",
        "isFork": true,
        "name": "mobx-state-tree",
        "stargazers": Object {
          "totalCount": 0,
        },
        "updatedAt": "2019-02-11T09:05:19Z",
      },
      Object {
        "createdAt": "2019-01-18T01:49:29Z",
        "description": "Automatic and enhanced Google Analytics tracking for common user interactions on the web.",
        "id": "MDEwOlJlcG9zaXRvcnkxNjYzMjU0NzI=",
        "isFork": true,
        "name": "autotrack",
        "stargazers": Object {
          "totalCount": 0,
        },
        "updatedAt": "2019-01-18T01:49:32Z",
      },
      Object {
        "createdAt": "2018-12-09T05:00:34Z",
        "description": "üë©‚Äçüíª üë®‚Äçüíª This is a repository for developing pyconkr frontend.",
        "id": "MDEwOlJlcG9zaXRvcnkxNjEwMDU5NDM=",
        "isFork": false,
        "name": "pyconkr-web",
        "stargazers": Object {
          "totalCount": 18,
        },
        "updatedAt": "2019-06-11T09:47:46Z",
      },
      Object {
        "createdAt": "2018-10-07T07:46:13Z",
        "description": "Dash and Zeal documentation integration in Visual Studio Code",
        "id": "MDEwOlJlcG9zaXRvcnkxNTE5MTc4NDk=",
        "isFork": true,
        "name": "vscode-dash",
        "stargazers": Object {
          "totalCount": 0,
        },
        "updatedAt": "2018-10-25T10:12:53Z",
      },
    ],
  },
}
`;
